README
AiChaos Alpha Chatterbot Module
-------------------------------

Installation:

	Extract the modules to your perl\lib\Chatbot folder.

Usage:

	See the POD data within Alpha.pm.

Example:

	This distribution comes with a test.pl (for testing) and a
	testreply.txt (a simple reply file that tests all the features
	of Alpha).

-------------------------------

Copyright (C) 2005 Cerone Kirsle
cjkirsle[at]aichaos[dot]com

Changes
Revision history for Perl extension Chatbot::Alpha.

2.04  Mon Dec 19 15:42:00
	- Fixed up some Perl warnings within the code.
	- Renamed the example script to 'example.pl' to not confuse Makefile.

2.03  Sat Dec 17 12:17:00 2005
	- Added ~REGEXP command.
	- Added {random} tag.
	- Added more information about #CODE on manpage.
	- Updated Chatbot::Alpha::Syntax to support the new ~REGEXP command.
	- Applied a patch to Syntax.pm to hopefully make automatic installation
		run more smoothly.

2.02  Thu Oct 27 10:57:30 2005
	- Mostly bug fixes in this release:
	- Added 'verify' argument to the new() constructor. If 1 (default),
		Chatbot::Alpha::Syntax is run on all files loaded. Set to 0 to avoid
		syntax checking (not recommended).
	- Fixed regexp bug with {formal}, {sentence}, {uppercase}, and {lowercase}.
		They should now function correctly even if their values have odd characters
		in them that would've previously screwed up the regexp parser.
	- Chatbot::Alpha::Syntax updated. See its manpage for details.
	- Rearranged a bit of the code so that <input> and <reply> would process
		before string tags.

2.01  Thu Oct 13 15:23:00 2005
	- Added string tags {formal}, {sentence}, {uppercase}, {lowercase}
	- Added tags <input> and <reply> and alias <star>.
	- Fixed conditionals bug (conditionals wouldn't increment correctly so
		only the last condition remained in memory. Now multiple conditions
		can be used for one trigger... i.e. comparing gender to male/female
		in two different conditions).

2.00  Thu Oct  6 18:53:00 2005
	- Added some AIML emulation:
		- In-reply redirections (like <srai>):
			+ * or something
			- Or something. {@<star1>}
	- "That" kind of support.
		+ i hate you
		- You're really mean... =(

		+ sorry
		% youre really mean
		- Don't worry--it's okay. :-)
	- Renamed all methods to be alternatingCaps instead of with underscores.
	- Chatbot::Alpha::Syntax supports the newly-added commands.
	- Fixed conditionals, should work more efficiently now:
		- Format changed to *VARIABLE=VALUE::HAPPENS
		- Does numeric == for numbers, or eq for strings... = better matching.

1.71  Sat Oct  1 14:01:00 2005
	- Sorting method reprogrammed for more efficient sorting.

1.70  Wed Aug 17 13:47:35 2005
        - Chatbot::Alpha::Syntax added.
        - ^ command added.
        - Module keeps filenames and line numbers internally, so on internal
                errors such as 'Deep Recursion' and 'Infinite Loop' it can point you
                to the source of the problem.
        - $alpha->search() method added.

1.61  Sat Aug  6 14:43:00 2005
        - Completed Chatbot::Alpha::Sort module.

1.40  Wed May  4 14:54:00 2005
	- Fixed major bug: * being converted to (.*?) wasn't a global replace, so
		thus far reply triggers could only contain a single *.

1.30  Tue Feb 15 16:12:00 2004
	- Added capability for topics.

1.10  Thu Dec 23 14:20:23 2004
	- Added the "#" command.
	- Fixed bugs in reply matching.

1.00  Wed Sep  6 16:22:40 2004
	- original version; created by h2xs 1.23 with options
		-X -n Chatbot::Alpha



test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..1\n"; }
END {print "not ok 1\n" unless $loaded;}
use Chatbot::Alpha;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

example.pl
#!/usr/bin/perl -w

use lib "./lib";
use Data::Dumper;
use Chatbot::Alpha;

my $alpha = new Chatbot::Alpha (debug => 1);
print "Chatbot::Alpha version $Chatbot::Alpha::VERSION\n\n";

# Load test replies.
my $load = $alpha->loadFile ('./testreplies.txt');
die "Error: $load" unless $load == 1;

print "\n\n";

# Stream additional replies.
$alpha->stream ("+ what is alpha\n"
	. "- Alpha, aka Chatbot::Alpha, is a chatterbot brain created by AiChaos Inc.\n\n"
	. "+ who created alpha\n"
	. "- Chatbot::Alpha was created by Cerone Kirsle.");

print "\n\n\n\n";

$alpha->sortReplies;

# User ID (so the module can keep track of different talkers)
my $id = "foo";

print "Setting Variables Test\n"
	. "Botmaster? [1|0] or <0> ";
my $bm = <STDIN>;
print "Name? or <user> ";
my $name = <STDIN>;

$bm ||= 0;
$name ||= 'user';

chomp $bm;
chomp $name;
$name = lc($name);

$alpha->setVariable ("master",$bm);
$alpha->setVariable ("name",$name);

# Loop.
while (1) {
	print "  You> ";
	my $msg = <STDIN>;
	chomp $msg;

	exit(0) if $msg =~ /^exit$/i;

	# Get a reply.
	my $reply = $alpha->reply ($id,$msg);

	$reply =~ s/\\n/\n/g;

	print "Alpha> $reply\n\n";
}

lib/Chatbot/Alpha/Syntax.pm
package Chatbot::Alpha::Syntax;

our $VERSION = '0.3';

use strict;
use warnings;

sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto;

	my $self = {
		debug    => 0,
		version  => $VERSION,
		deny     => {},
		allow    => {},
		denytype => 'alloy_all', # deny_all, allow_some, deny_some
		cusdeny  => 0,
		syntax   => 'strict',
	};

	bless ($self,$class);
	return $self;
}

sub syntax {
	my ($self,$syn) = @_;

	if ($syn =~ /^(loose|strict)$/i) {
		$self->{syntax} = $syn;
		return 1;
	}

	return 0;
}

sub deny_type {
	my ($self,$type) = @_;

	if ($type =~ /^(alloy|deny)_(all|some)$/i) {
		$self->{cusdeny} = 1;

		$type = lc($type);
		$type =~ s/ //g;

		$self->{denytype} = $type;
	}
	else {
		return 0;
	}
	return 1;
}

sub deny {
	my ($self,@commands) = @_;

	# Deny each command.
	foreach my $cmd (@commands) {
		delete $self->{allow}->{$cmd} if exists $self->{allow}->{$cmd};
		$self->{deny}->{$cmd} = 1;
	}

	$self->deny_type ('deny_some') unless $self->{cusdeny} == 1;
}

sub allow {
	my ($self,@commands) = @_;

	# Allow each command.
	foreach my $cmd (@commands) {
		delete $self->{deny}->{$cmd} if exists $self->{deny}->{$cmd};
		$self->{allow}->{$cmd} = 1;
	}

	$self->deny_type ('allow_some') unless $self->{cusdeny} == 1;
}

sub check {
	my ($self,$file) = @_;

	open (FILE, $file) or return 0;
	my @data = <FILE>;
	close (FILE);

	# Handle dos text files on Mac and Unix
	if($/ ne "\r\n") {
		local $/ = "\r\n";
		chomp @data;
	}

	chomp @data;

	# Go through each line.
	my $num = 0;
	foreach my $line (@data) {
		$num++;
		next if length $line == 0;
		next if $line =~ /^\//;
		$line =~ s/^\s+//g;
		$line =~ s/^\t+//g;
		$line =~ s/^\s//g;
		$line =~ s/^\t//g;

		my ($cmd,$data) = split(//, $line, 2);
		$data =~ s/^\s+//g;
		$data =~ s/^\s//g;

		next unless length $cmd > 0;

		# Denied/Not allowed?
		if ($self->{denytype} ne 'allow_all') {
			if ($self->{denytype} eq 'deny_some') {
				if (exists $self->{deny}->{$cmd}) {
					die "Command $cmd is not allowed at $file line $num; ";
				}
			}
			elsif ($self->{denytype} eq 'allow_some') {
				if (!exists $self->{allow}->{$cmd}) {
					die "Command $cmd not in allowlist at $file line $num; ";
				}
			}
		}
		elsif ($self->{denytype} eq 'deny_all') {
			die "No commands allowed at $file line $num; ";
		}

		if ($cmd eq '>') {
			my @args = split(/\s+/, $data);
			if (scalar(@args) != 2) {
				die "Bad number of arguments in >LABEL at $file line $num; ";
			}
		}
		elsif ($cmd eq '<') {
			my @args = split(/\s+/, $data);
			if (scalar(@args) != 1) {
				die "Bad number of arguments in <LABEL at $file line $num; ";
			}
		}
		elsif ($cmd eq '+') {
			# On strict: must be lowercase, simplistic.
			if ($self->{syntax} eq 'strict') {
				if ($data =~ /[^a-z0-9 \*]/) {
					die "+TRIGGERS must be lowercase alphanumeric "
						. "while in 'strict' syntax at $file line $num; ";
				}
			}
			elsif ($self->{syntax} eq 'loose') {
				if ($data =~ /[^A-Za-z0-9 \*]/) {
					warn "+TRIGGERS must be alphanumeric while in 'loose' "
						. "syntax at $file line $num; ";
				}
			}
		}
		elsif ($cmd eq '%') {
			# On strict: must be lowercase, simplistic.
			if ($self->{syntax} eq 'strict') {
				if ($data =~ /[^a-z0-9 ]/) {
					die "+TRIGGERS must be lowercase alphanumeric "
						. "while in 'strict' syntax at $file line $num; ";
				}
			}
			elsif ($self->{syntax} eq 'loose') {
				if ($data =~ /[^A-Za-z0-9 ]/) {
					warn "+TRIGGERS must be alphanumeric while in 'loose' "
						. "syntax at $file line $num; ";
				}
			}
		}
		elsif ($cmd eq '-') {
			if (length $data == 0) {
				die "Empty -RESPONSE data at $file line $num; ";
			}
		}
		elsif ($cmd eq '^') {
			if (length $data == 0) {
				die "Empty ^CONTINUE data at $file line $num; ";
			}
		}
		elsif ($cmd eq '@') {
			if ($self->{syntax} eq 'strict') {
				if ($data =~ /[^a-z0-9 \*\<\>]/) {
					die "\@REDIRECTIONS must be lowercase alphanumeric "
						. "while in 'strict' syntax at $file line $num; ";
				}
			}
			elsif ($self->{syntax} eq 'loose') {
				if ($data =~ /[^A-Za-z0-9 \*\<\>]/) {
					die "\@REDIRECTIONS must be alphanumeric while in 'loose' "
						. "syntax at $file line $num; ";
				}
			}
		}
		elsif ($cmd eq '*') {
			if ($data !~ /^(.*?)=(.*?)::(.*?)$/i) {
				die "Syntax error at *CONDITION at $file line $num; ";
			}
		}
		elsif ($cmd eq '&') {
			if (length $data == 0) {
				die "Empty &HOLDER data at $file line $num; ";
			}
		}
		elsif ($cmd eq '#') {
			if (length $data == 0) {
				die "Empty #CODE data at $file line $num; ";
			}
		}
		elsif ($cmd eq '/') {
			# Comment data.
		}
		elsif ($cmd eq '~') {
			# A regexp. Leave it be.
		}
		else {
			warn "Unknown command '$cmd' with data '$data' at $file line $num; ";
		}
	}

	return 1;
}

1;
__END__

=head1 NAME

Chatbot::Alpha::Syntax - Syntax checking for Chatbot::Alpha replies.

=head1 SYNOPSIS

  use Chatbot::Alpha::Syntax;
  
  my $syntax = new Chatbot::Alpha::Syntax;
  
  # Set 'strict' syntax.
  $syntax->syntax ('strict');
  
  # Changed my mind, use 'loose'
  $syntax->syntax ('loose');
  
  # Only allow SOME commands.
  $syntax->deny_type ('allow_some');
  
  # Allow only +'s and -'s.
  $syntax->allow ('+', '-');
  
  # Syntax-check this file.
  $syntax->check ("replies.cba");

=head1 DESCRIPTION

Chatbot::Alpha::Syntax provides syntax checking for Alpha documents. All syntax errors
result in a 'die' so don't expect to run your syntax checking halfway through a large
application's process. Doing it in initialization is always fine though.

=head1 METHODS

=head2 new (ARGUMENTS)

Creates a new Chatbot::Alpha::Syntax object. You can pass in any defaults here.

=head2 syntax (TYPE)

Define a syntax type, either 'strict' or 'loose'. Defaults to strict. See below for definitions
on the various syntax types.

=head2 deny_type (DENYTYPE)

Must be 'deny_all', 'deny_some', 'allow_some', or 'allow_all' - defaults to 'allow_all'. If you're
going to want to deny/allow certain commands, it's best to use deny_type to set this. The automatic
settings of deny() and allow() may not always end up how you want them.

=head2 deny (COMMANDS)

Denies a list of COMMANDS. These are the Alpha commands (+, -, @, &, etc). Syntax errors will
arrise when these commands are found in the Alpha document.

=head2 allow (COMMANDS)

Adds COMMANDS to the allow list.

=head2 check (FILE)

Check the syntax of FILE. Will return 0 if the file couldn't be opened, return 1 if everything
went well, or die if a syntax error is found.

=head1 SYNTAX TYPES

Syntax types mostly only refer to the +TRIGGER command, as that's the part of your code that's
put through a regexp.

=head2 strict

This is the default (and most recommended) syntax type. The rules are as follows:

  - Triggers must be lowercase, numbers and letters only.
  - Spaces are allowed. All other symbols are NOT allowed.

=head2 loose

This one is less strict on your trigger syntax. The recommended rules are as follows:

  - Triggers can be capitilized, lowercase, or any combination.
  - Triggers can contain letters or numbers or spaces.
  - Any foreign symbols aren't recommended, however it won't kill you.

The loose syntax check will only 'warn' when one of these isn't true, but it won't
hold it against you.

=head1 ALPHA SYNTAX

Here is the proper syntax of each Alpha command.

=head2 +TRIGGER

See SYNTAX TYPES.

=head2 ~REGEXP

No syntax rules have been applied to these. Just make sure your regexp triggers are
friendly.

=head2 -RESPONSE

A value of any length must be given. A response of all spaces is bad.

=head2 >LABEL

Two arguments must be given, separated by spaces: the label type, and its
one-word value.

=head2 <LABEL

One argument given.

=head2 @REDIRECT

Follows the same rules as +TRIGGER

=head2 &HOLDER

Follows the same rules as -RESPONSE

=head2 *CONDITION

Must follow this syntax exactly:

  * ___=___::___
    ^var  ^val ^response

=head2 #CODE

Must have a length to it.

=head1 KNOWN BUGS

No bugs known at the moment.

=head1 CHANGES

  Version 0.2
  - Fixed some bugs, blank lines shouldn't ever be considered commands,
    and incase of unknown command anyway only a warn is used but not a
    die.

  Version 0.1
  - Initial release.

=head1 FUTURE PLANS

  - Add methods for defining your own syntax, for example if you make
    a custom mod to Chatbot::Alpha to add new commands, the syntax
    checker would know what to do with them.

=head1 SEE ALSO

L<Chatbot::Alpha>

=head1 AUTHOR

Cerone J. Kirsle, cjkirsle "@" aichaos.com

=head1 COPYRIGHT AND LICENSE

    Chatbot::Alpha - A simple chatterbot brain.
    Copyright (C) 2005  Cerone J. Kirsle

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=cut

lib/Chatbot/Alpha/Sort.pm
package Chatbot::Alpha::Sort;

our $VERSION = '0.2';

use strict;
use warnings;
use Chatbot::Alpha;

sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto || 'Chatbot::Alpha::Sort';

	my $self = {
		debug   => 0,
		version => $VERSION,
		@_,
	};

	bless ($self,$class);

	return $self;
}

sub debug {
	my ($self,$msg) = @_;

	# Only show if debug mode is on.
	if ($self->{debug} == 1) {
		print STDOUT "Alpha::Sort::Debug // $msg\n";
	}

	return 1;
}

sub start {
	my $self = shift;

	# $sort->start (
	#    dir   => './my_replies',
	#    files => 'alpha|intact|single',
	#    ext   => 'cba',
	# );
	#  alpha: A.cba to Z.cba, star.cba and other.cba
	# intact: keeps filenames the same, sorts replies internally
	# single: all replies to a single file.

	# Defaults.
	my $method = {
		dir   => '.',
		out   => '.',
		files => 'alpha',
		ext   => 'cba',
		@_,
	};

	# Make outdir if it doesn't exist.
	if ($method->{out} ne '.' && !-e $method->{out}) {
		mkdir ($method->{out});
	}

	# Alpha sorting: sorts from A.cba to Z.cba
	if ($method->{files} eq 'alpha') {
		$self->debug ("Sorting files into alphabetics.");

		# Load ALL replies.
		my $alpha = new Chatbot::Alpha();
		$alpha->load_folder ($method->{dir}, $method->{ext});

		my @labs = qw(A B C D E F G H I J K L M N O P Q R S T U V W X Y Z numbers topics star other);
		foreach my $lab (@labs) {
			$self->{lab}->{$lab} = [];
		}

		# Get an array of all triggers and then sort.
		my @top = keys %{$alpha->{_replies}->{random}};
		@top = sort (@top);

		$self->debug ("Going through triggers...");
		foreach my $trig (@top) {
			$trig =~ s/\(\.\*\?\)/\*/ig;
			$self->debug ("\tTrigger: $trig");
			my $first = substr ($trig,0,1);
			$first = uc($first);
			$first = 'star' if $first eq '*';
			$self->debug ("\t\tFirst: $first");
			if (exists $self->{lab}->{$first}) {
				push (@{$self->{lab}->{$first}}, $trig);
			}
			else {
				push (@{$self->{lab}->{other}}, $trig);
			}
		}

		# Write to files.
		foreach my $lab (keys %{$self->{lab}}) {
			open (WRITE, ">$method->{out}/$lab\.$method->{ext}");

			foreach my $trig (@{$self->{lab}->{$lab}}) {
				$self->debug ("Writing trigger $trig to $lab\.$method->{ext}");

				print WRITE "+ $trig\n";

				$trig =~ s/\*/\(\.\*\?\)/ig;

				my @data = keys (%{$alpha->{_replies}->{random}->{$trig}});
				@data = sort(@data);

				foreach my $item (@data) {
					if ($item =~ /^\d/) {
						print WRITE "- $alpha->{_replies}->{random}->{$trig}->{$item}\n";
					}
					elsif ($item =~ /^(conditions|convo)$/i) {
						my @sub = keys %{$alpha->{_replies}->{random}->{$trig}->{$item}};
						@sub = reverse(@sub);
						foreach my $s (@sub) {
							if ($item eq 'conditions') {
								print WRITE "* ";
							}
							else {
								print WRITE "& ";
							}
							print WRITE "$alpha->{_replies}->{random}->{$trig}->{$item}->{$s}\n";
						}
					}
					elsif ($item eq 'redirect') {
						print WRITE "\@ $alpha->{_replies}->{random}->{$trig}->{redirect}\n";
					}
					elsif ($item eq 'system') {
						print WRITE "# $alpha->{_replies}->{random}->{$trig}->{system}->{codes}\n";
					}
				}

				print WRITE "\n";
			}

			close (WRITE);
		}

		# Write to topics.
		open (TOPICS, ">$method->{out}/topics\.$method->{ext}");
		foreach my $topic (keys %{$alpha->{_replies}}) {
			next if $topic eq 'random';

			my @trigs = keys %{$alpha->{_replies}->{$topic}};
			@trigs = sort(@trigs);

			print TOPICS "> topic $topic\n\n";

			foreach my $trig (@trigs) {
				$self->debug ("Writing trigger $trig to topics\.$method->{ext}");

				$trig =~ s/\(\.\*\?\)/\*/ig;

				print TOPICS "\t+ $trig\n";

				$trig =~ s/\*/\(\.\*\?\)/ig;

				my @data = keys (%{$alpha->{_replies}->{$topic}->{$trig}});
				@data = sort(@data);

				foreach my $item (@data) {
					if ($item =~ /^\d/) {
						print TOPICS "\t- $alpha->{_replies}->{$topic}->{$trig}->{$item}\n";
					}
					elsif ($item =~ /^(conditions|convo)$/i) {
						my @sub = keys %{$alpha->{_replies}->{$topic}->{$trig}->{$item}};
						@sub = reverse(@sub);
						foreach my $s (@sub) {
							if ($item eq 'conditions') {
								print TOPICS "\t* ";
							}
							else {
								print TOPICS "\t& ";
							}
							print TOPICS "$alpha->{_replies}->{$topic}->{$trig}->{$item}->{$s}\n";
						}
					}
					elsif ($item eq 'redirect') {
						print TOPICS "\t\@ $alpha->{_replies}->{$topic}->{$trig}->{redirect}\n";
					}
					elsif ($item eq 'system') {
						print TOPICS "\t# $alpha->{_replies}->{$topic}->{$trig}->{system}->{codes}\n";
					}
				}

				print TOPICS "\n";
			}

			print TOPICS "< topic\n\n";
		}
		close (TOPICS);

		return 1;
	}
	elsif ($method->{files} eq 'intact') {
		# Keeping files intact.
		$self->debug ("Keeping filenames intact");

		opendir (DIR, "$method->{dir}");
		foreach my $file (sort(grep(!/^\./, readdir(DIR)))) {
			next unless $file =~ /\.$method->{ext}/i;

			# Create a new Alpha object for this file.
			my $alpha = new Chatbot::Alpha();
			$alpha->load_file ("$method->{dir}/$file");

			open (WRITE, ">$method->{out}/$file");

			# Do topics first.
			foreach my $topic (keys %{$alpha->{_replies}}) {
				next if $topic eq 'random';
				my @trigs = keys %{$alpha->{_replies}->{$topic}};
				@trigs = sort(@trigs);

				print WRITE "> topic $topic\n\n";

				foreach my $trig (@trigs) {
					$self->debug ("Writing trigger $trig to topics\.$method->{ext}");

					$trig =~ s/\(\.\*\?\)/\*/ig;

					print WRITE "\t+ $trig\n";

					$trig =~ s/\*/\(\.\*\?\)/ig;

					my @data = keys (%{$alpha->{_replies}->{$topic}->{$trig}});
					@data = sort(@data);

					foreach my $item (@data) {
						if ($item =~ /^\d/) {
							print WRITE "\t- $alpha->{_replies}->{$topic}->{$trig}->{$item}\n";
						}
						elsif ($item =~ /^(conditions|convo)$/i) {
							my @sub = keys %{$alpha->{_replies}->{$topic}->{$trig}->{$item}};
							@sub = reverse(@sub);
							foreach my $s (@sub) {
								if ($item eq 'conditions') {
									print WRITE "\t* ";
								}
								else {
									print WRITE "\t& ";
								}
								print WRITE "$alpha->{_replies}->{$topic}->{$trig}->{$item}->{$s}\n";
							}
						}
						elsif ($item eq 'redirect') {
							print WRITE "\t\@ $alpha->{_replies}->{$topic}->{$trig}->{redirect}\n";
						}
						elsif ($item eq 'system') {
							print WRITE "\t# $alpha->{_replies}->{$topic}->{$trig}->{system}->{codes}\n";
						}
					}

					print WRITE "\n";
				}

				print WRITE "< topic\n\n";
			}

			# Now, normal replies.
			my @trigs = keys %{$alpha->{_replies}->{random}};
			@trigs = sort(@trigs);

			foreach my $trig (@trigs) {
				$self->debug ("Writing trigger $trig to topics\.$method->{ext}");

				$trig =~ s/\(\.\*\?\)/\*/ig;

				print WRITE "+ $trig\n";

				$trig =~ s/\*/\(\.\*\?\)/ig;

				my @data = keys (%{$alpha->{_replies}->{random}->{$trig}});
				@data = sort(@data);

				foreach my $item (@data) {
					if ($item =~ /^\d/) {
						print WRITE "- $alpha->{_replies}->{random}->{$trig}->{$item}\n";
					}
					elsif ($item =~ /^(conditions|convo)$/i) {
						my @sub = keys %{$alpha->{_replies}->{random}->{$trig}->{$item}};
						@sub = reverse(@sub);
						foreach my $s (@sub) {
							if ($item eq 'conditions') {
								print WRITE "* ";
							}
							else {
								print WRITE "& ";
							}
							print WRITE "$alpha->{_replies}->{random}->{$trig}->{$item}->{$s}\n";
						}
					}
					elsif ($item eq 'redirect') {
						print WRITE "\@ $alpha->{_replies}->{random}->{$trig}->{redirect}\n";
					}
					elsif ($item eq 'system') {
						print WRITE "# $alpha->{_replies}->{random}->{$trig}->{system}->{codes}\n";
					}
				}

				print WRITE "\n";
			}

			close (WRITE);
		}
		closedir (DIR);

		return 1;
	}
	elsif ($method->{files} eq 'single') {
		$self->debug ("Merging all files into one");

		my $alpha = new Chatbot::Alpha();
		$alpha->load_folder ($method->{dir},$method->{ext});

		open (WRITE, ">$method->{out}/sorted.cba");

		# Do topics first.
		foreach my $topic (keys %{$alpha->{_replies}}) {
			next if $topic eq 'random';
			my @trigs = keys %{$alpha->{_replies}->{$topic}};
			@trigs = sort(@trigs);

			print WRITE "> topic $topic\n\n";

			foreach my $trig (@trigs) {
				$self->debug ("Writing trigger $trig to topics\.$method->{ext}");

				$trig =~ s/\(\.\*\?\)/\*/ig;

				print WRITE "\t+ $trig\n";

				$trig =~ s/\*/\(\.\*\?\)/ig;

				my @data = keys (%{$alpha->{_replies}->{$topic}->{$trig}});
				@data = sort(@data);

				foreach my $item (@data) {
					if ($item =~ /^\d/) {
						print WRITE "\t- $alpha->{_replies}->{$topic}->{$trig}->{$item}\n";
					}
					elsif ($item =~ /^(conditions|convo)$/i) {
						my @sub = keys %{$alpha->{_replies}->{$topic}->{$trig}->{$item}};
						@sub = reverse(@sub);
						foreach my $s (@sub) {
							if ($item eq 'conditions') {
								print WRITE "\t* ";
							}
							else {
								print WRITE "\t& ";
							}
							print WRITE "$alpha->{_replies}->{$topic}->{$trig}->{$item}->{$s}\n";
						}
					}
					elsif ($item eq 'redirect') {
						print WRITE "\t\@ $alpha->{_replies}->{$topic}->{$trig}->{redirect}\n";
					}
					elsif ($item eq 'system') {
						print WRITE "\t# $alpha->{_replies}->{$topic}->{$trig}->{system}->{codes}\n";
					}
				}

				print WRITE "\n";
			}

			print WRITE "< topic\n\n";
		}

		# Now, normal replies.
		my @trigs = keys %{$alpha->{_replies}->{random}};
		@trigs = sort(@trigs);

		foreach my $trig (@trigs) {
			$self->debug ("Writing trigger $trig to topics\.$method->{ext}");

			$trig =~ s/\(\.\*\?\)/\*/ig;

			print WRITE "+ $trig\n";

			$trig =~ s/\*/\(\.\*\?\)/ig;

			my @data = keys (%{$alpha->{_replies}->{random}->{$trig}});
			@data = sort(@data);

			foreach my $item (@data) {
				if ($item =~ /^\d/) {
					print WRITE "- $alpha->{_replies}->{random}->{$trig}->{$item}\n";
				}
				elsif ($item =~ /^(conditions|convo)$/i) {
					my @sub = keys %{$alpha->{_replies}->{random}->{$trig}->{$item}};
					@sub = reverse(@sub);
					foreach my $s (@sub) {
						if ($item eq 'conditions') {
							print WRITE "* ";
						}
						else {
							print WRITE "& ";
						}
						print WRITE "$alpha->{_replies}->{random}->{$trig}->{$item}->{$s}\n";
					}
				}
				elsif ($item eq 'redirect') {
					print WRITE "\@ $alpha->{_replies}->{random}->{$trig}->{redirect}\n";
				}
				elsif ($item eq 'system') {
					print WRITE "# $alpha->{_replies}->{random}->{$trig}->{system}->{codes}\n";
				}
			}

			print WRITE "\n";
		}

		close (WRITE);

		return 1;
	}
}

1;
__END__

=head1 NAME

Chatbot::Alpha::Sort - Alphabetic sorting for Chatbot::Alpha documents.

=head1 SYNOPSIS

  use Chatbot::Alpha::Sort;
  
  # Create a new sorter.
  my $sort = new Chatbot::Alpha::Sort();
  
  # Sort your files.
  $sort->start (
     dir => './before',
     out => './after',
     ext => 'cba',
  );

=head1 DESCRIPTION

Chatbot::Alpha::Sort can take your numerous unsorted Alpha documents, and create nicely formatted
documents from A.cba to Z.cba with triggers sorted alphabetically within each document.

=head1 METHODS

=head2 new (ARGUMENTS)

Creates a new Chatbot::Alpha::Sort object. You should only need one object, since each sort request
creates its own Chatbot::Alpha, unless you intend to run multiple sorts at the same time.

Returns a Chatbot::Alpha::Sort instance.

=head2 version

Returns the version number of the module.

=head2 start (ARGUMENTS)

Starts the sorting process. ARGUMENTS is a hash that you must pass in to tell the module how to do things. The arguments
are as follows:

  dir => DIRECTORY
     The directory at which your original Alpha documents
     can be found. Defaults to CWD.
  out => DIRECTORY
     Another directory for which your newly formatted Alpha
     documents will be written to. Defaults to CWD.
  ext => EXTENSION
     The file extension of your Alpha documents. Defaults
     to cba
  files => SORT_TYPE
     The sorting method for which your new files will be sorted.
     See below for the sort types.

=head1 SORT TYPES (FILES)

B<alpha>
Sorts the files alphabetically. Will create files "A.cba" through "Z.cba", as well as "star.cba" and "other.cba",
and a "topics.cba" to keep all topics together. Triggers within each file are sorted alphabetically.

B<intact>
Keeps your original filename structure intact; only sorts triggers alphabetically within each file (topics go
to the top of the file's contents).

B<single>
Takes ALL your Alpha documents and merges them into one single file. Will write the finished file to "sorted.cba"
in the OUT directory. Topics are written first.

=head1 CHANGES

  Version 0.2
  - Added sorting with "keep filenames intact" as well as "single filename"
  
  Version 0.1
  - Initial release.

=head2 SEE ALSO

L<Chatbot::Alpha>

=head1 KNOWN BUGS

No bugs have been discovered at this time.

=head1 AUTHOR

Cerone J. Kirsle, cjkirsle "@" aichaos.com

=head1 COPYRIGHT AND LICENSE

    Chatbot::Alpha - A simple chatterbot brain.
    Copyright (C) 2005  Cerone J. Kirsle

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=cut